packages/core-backend/src/BackendWebSocketService.ts

```diff
@@ -0,0 +1,1316 @@
+        console.debug(
+          `[${SERVICE_NAME}] Reconnection disabled by enabledCallback (app closed/backgrounded) - stopping all reconnection attempts`,
+        );
+        this.#reconnectAttempts = 0;
+        return;
+      }
+
+      // Authentication checks are handled in connect() method
+      // No need to check here since AuthenticationController manages wallet state internally
+
+      console.debug(
+        `🔄 ${delay}ms delay elapsed - starting reconnection attempt #${this.#reconnectAttempts}...`,
+      );
+
+      this.connect().catch((error) => {
+        console.error(
+          `❌ Reconnection attempt #${this.#reconnectAttempts} failed:`,
+          error,
+        );
+
+        // Always schedule another reconnection attempt
+        this.#scheduleReconnect();
+      });
+    }, delay);
+  }
+
+  /**
+   * Clears all active timers
+   */
+  #clearTimers(): void {
+    if (this.#reconnectTimer) {
+      clearTimeout(this.#reconnectTimer);
+      this.#reconnectTimer = null;
+    }
+  }
+
+  /**
+   * Clears all pending requests and rejects them with the given error
+   *
+   * @param error - Error to reject with
+   */
+  #clearPendingRequests(error: Error): void {
+    for (const [, request] of this.#pendingRequests) {
+      clearTimeout(request.timeout);
+      request.reject(error);
+    }
+    this.#pendingRequests.clear();
+  }
+
+  /**
+   * Clears all active subscriptions
+   */
+  #clearSubscriptions(): void {
+    this.#subscriptions.clear();
+  }
+
+  /**
+   * Sets the connection state and emits state change events
+   *
+   * @param newState - The new WebSocket state
+   */
+  #setState(newState: WebSocketState): void {
+    const oldState = this.#state;
+    this.#state = newState;
+
+    if (oldState !== newState) {
+      console.debug(`WebSocket state changed: ${oldState} → ${newState}`);
+
+      // Log disconnection-related state changes
+      if (
+        newState === WebSocketState.DISCONNECTED ||
+        newState === WebSocketState.DISCONNECTING ||
+        newState === WebSocketState.ERROR
+      ) {
+        console.debug(
+          `🔴 WebSocket disconnection detected - state: ${oldState} → ${newState}`,
+        );
+      }
+
+      // Publish connection state change event
+      try {
+        this.#messenger.publish(
+          'BackendWebSocketService:connectionStateChanged',
+          this.getConnectionInfo(),
+        );
+      } catch (error) {
+        console.error(
+          'Failed to publish WebSocket connection state change:',
+          error,
+        );
+      }
+    }
+  }
+
+  // =============================================================================
+  // 7. UTILITY METHODS (PRIVATE)
+  // =============================================================================
+
+  /**
+   * Extracts error message from unknown error type
+   *
+   * @param error - Error of unknown type
+   * @returns Error message string
+   */
+  #getErrorMessage(error: unknown): string {
```

Comment:
The `getErrorMessage` function from `@metamask/utils` does this already, if I'm not mistaken?